src/pj_obs_api.c

```diff
@@ -139,147 +121,306 @@ PJ_OBS pj_trans (PJ *P, enum pj_direction direction, PJ_OBS obs) {
-    PJ_CONTEXT *ctx;
-    if (0==P)
-        ctx = pj_get_default_ctx();
-    else
-        ctx = pj_get_ctx (P);
-    previous = ctx->debug_level;
-    if (PJ_LOG_TELL==log_level)
-        return previous;
-    ctx->debug_level = log_level;
-    return previous;
-}
+    Has a synonym defined in proj.h: proj_errno_restore (P, err), but perhaps
+    the naming should be the other way round, since the errno reset/restore
+    paradigm will probably be the primary mode of operation.
 
+    See usage example under proj_errno_reset ()
+******************************************************************************/
+    if (0==P) {
+        errno = EINVAL;
+        return;
+    }
+
+    /* Use proj_errno_reset to explicitly clear the error status */
+    if (0==err)
+        return;
 
-/* Put a new logging function into P's context. The opaque object app_data is passed as first arg at each call to the logger */
-void pj_log_func (PJ *P, void *app_data, void (*log)(void *, int, const char *)) {
-    PJ_CONTEXT *ctx = pj_get_ctx (P);
-    ctx->app_data = app_data;
-    if (0!=log)
-        ctx->logger = log;
+    /* set local error level */
+    P->last_errno = err;
+    /* and let it bubble up */
+    proj_context_errno_set (pj_get_ctx (P), err);
+    errno = err;
+    return;
 }
 
 
-/* Move P to a new context - initially a copy of the default context */
-int pj_context_renew (PJ *P) {
-    PJ_CONTEXT *ctx = pj_ctx_alloc ();
-    if (0==ctx) {
-        pj_err_level (P, ENOMEM);
-        return 1;
+/*****************************************************************************/
+int proj_errno_reset (PJ *P) {
+/******************************************************************************
+    Clears errno in the PJ, and bubbles it up to the context and
+    pj_errno levels through the low level pj_ctx interface.
+    
+    Returns the previous value of the errno, for convenient reset/restore
+    operations:
+
+    void foo (PJ *P) {
+        int last_errno = proj_errno_reset (P);
+        
+        do_something_with_P (P);
+
+        (* failure - keep latest error status *)
+        if (proj_errno(P))
+            return;
+        (* restore previous error status *)
+        proj_errno_restore (P, last_errno);
+        return;
     }
-
-    pj_set_ctx (P, ctx);
-    return 0;
+******************************************************************************/
+    int last_errno;
+    if (0==P) {
+        errno = EINVAL;
+        return EINVAL;
+    }
+    last_errno = proj_errno (P);
+
+    /* set local error level */
+    P->last_errno = 0;
+    /* and let it bubble up */
+    pj_ctx_set_errno (pj_get_ctx (P), 0);
+    errno = 0;
+    return last_errno;
 }
 
 
-/* Move daughter to mother's context */
-void pj_context_inherit (PJ *mother, PJ *daughter) {
-    pj_set_ctx (daughter, pj_get_ctx (mother));
+/* Create a new context - or provide a pointer to the default context */
+PJ_CONTEXT *proj_context_create (int multithreaded) {
```

Comment:
Wouldn't it be simpler to leave out the `multithreaded` argument? If you want the default context you call `proj_create(0, "...")` and if you want a non-default context you create it be calling `proj_context_create()`. In case you want to prepare your application for multithreaded use, but is still running it singlethreaded a non-default context can be used with out problem:

```
PJ_CONTEXT * ctx proj_context_create();
PJ * P proj_create(ctx, "+proj=...");
...
proj_context_destroy(ctx);
proj_destroy(P);
```

To me that is a little bit cleaner and simpler.